plugins/woocommerce/includes/data-stores/class-wc-customer-data-store-session.php

```diff
@@ -81,19 +84,41 @@ public function save_to_session( $customer ) {
 			if ( 'billing_' === substr( $session_key, 0, 8 ) ) {
 				$session_key = str_replace( 'billing_', '', $session_key );
 			}
-			$data[ $session_key ] = (string) $customer->{"get_$function_key"}( 'edit' );
+			if ( 'meta_data' === $session_key ) {
+				// This is an allow-list of meta data keys which we want to store in session. Avoids storing everything for logged in users.
+				$allowed_meta_data_keys = array(
+					CheckoutFields::ADDITIONAL_FIELDS_KEY,
+					CheckoutFields::BILLING_FIELDS_KEY,
+					CheckoutFields::SHIPPING_FIELDS_KEY,
+				);
+				$session_value          = wp_json_encode(
+					array_filter(
+						$customer->get_meta_data(),
+						function( $meta_data ) use ( $allowed_meta_data_keys ) {
+							return in_array( $meta_data->key, $allowed_meta_data_keys, true );
+						}
+					)
+				);
+			} else {
+				$session_value = $customer->{"get_$function_key"}( 'edit' );
+			}
+			$data[ $session_key ] = (string) $session_value;
 		}
 		WC()->session->set( 'customer', $data );
 	}
 
 	/**
-	 * Read customer data from the session unless the user has logged in, in
-	 * which case the stored ID will differ from the actual ID.
+	 * Read customer data from the session unless the user has logged in, in which case the stored ID will differ from
+	 * the actual ID.
 	 *
 	 * @since 3.0.0
 	 * @param WC_Customer $customer Customer object.
 	 */
 	public function read( &$customer ) {
+		if ( $customer->get_id() ) {
+			parent::read( $customer );
```

Comment:
This was added because we'll only load one data store from now on instead of both, when using sessions.